### PR TITLE
feat: unmanaged dependency check

### PR DESCRIPTION
**Pull request by JoeWang1127**
Implement an unmanaged dependency check, part of dependency governance project.

### CODE REVIEW

**Comment by suztomo** [comment]
Todo:

- To add action.yml a BOM directory is an input.
- Use the action.yml against https://github.com/googleapis/sdk-platform-java/tree/main/gapic-generator-java-bom
- Add an unmanaged dependency to https://github.com/googleapis/sdk-platform-java/blob/main/java-core/google-cloud-core/pom.xml to see the check fails (within this pull request).

Until the design doc is approved, leave this pull request as failed state in draft.

**Comment by JoeWang1127** [comment]
In order to fail the check, an unmanaged dependency has to be added in the target bom, not the dependency of this bom. 
In this case, an unmananged dependency has to be added in `gapic-generator-java-bom/pom.xml`, but not in `gax-java/gax-grpc/pom.xml`, which is a transitive dependency of `gapic-generator-java-bom/pom.xml`.

I added a dependency in `gapic-generator-java-bom/pom.xml` but the check is successful (it shouldn't)
```
Run unmanaged_dependencies=$(mvn exec:java -Dexec.args="3.20.0 ../../gapic-generator-java-bom/pom.xml" -q)
  unmanaged_dependencies=$(mvn exec:java -Dexec.args="3.20.0 ../../gapic-generator-java-bom/pom.xml" -q)
  echo "${unmanaged_dependencies}"
  if [[ "${unmanaged_dependencies}" != "[]" ]]; then
    echo "${unmanaged_dependencies}"
    exit 1
  fi
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    JAVA_HOME: /opt/hostedtoolcache/Java_Temurin-Hotspot_jdk/11.0.21-9/x64
    JAVA_HOME_11_X64: /opt/hostedtoolcache/Java_Temurin-Hotspot_jdk/11.0.21-9/x64
    shared_dependencies_version: 3.20.0
[]
```
When tested locally:
```
joewa-macbookpro:unmanaged-dependency-check joewa$ echo "${unmanaged_dependencies}"
[com.h2database:h2]
```

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -277,3 +277,21 @@ jobs:
...
+        java-version: 11
+        distribution: temurin
+    - run: mvn -version
+    - name: Install Maven modules
+      run: |
+        mvn install -B -ntp -DskipTests -Dclirr.skip -Dcheckstyle.skip
+    - name: Unmanaged dependency check
+      uses: ./java-shared-dependencies/unmanaged-dependency-check
+      with:
+        bom-path: ../../gapic-generator-java-bom/pom.xml
```
Comment (suztomo) [inline .github/workflows/ci.yaml:297]: This path needs to be relative to the repository root. The users (e.g., Java-storage) don’t
Know “../../” directory structure.

**Comment by suztomo** [comment]
> an unmanaged dependency has to be added in the target bom, not the dependency of this bom.

No, this check needs to detect unmanaged dependencies (accidentally) added as a dependency of a library in Cloud SDK for Java. A BOM as the argument is just convenience to specify the libraries to check.

**Review by suztomo — DISMISSED** [review]
unitTests (21) passed.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (suztomo) [reply .github/workflows/ci.yaml:297]: Check how action_path helps you https://docs.github.com/en/actions/learn-github-actions/contexts#github-context.

File: java-shared-dependencies/unmanaged-dependency-check/action.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+  name: "Unmanaged dependency check"
+  description: "Validation for "
```
Comment (suztomo) [inline java-shared-dependencies/unmanaged-dependency-check/action.yaml:2]: for ?

File: java-shared-dependencies/unmanaged-dependency-check/pom.xml [hunk 3]
```diff
@@ -0,0 +1,85 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>com.google.cloud</groupId>
+  <artifactId>unmanaged-dependency-check</artifactId>
+  <version>3.18.1-SNAPSHOT</version><!-- {x-version-update:google-cloud-shared-dependencies:current} -->
```
Comment (suztomo) [inline java-shared-dependencies/unmanaged-dependency-check/pom.xml:8]: Let's 0.0.1-SNAPSHOT and remove the release please annotation to ensure that we don't publish this artifact.

File: java-shared-dependencies/unmanaged-dependency-check/src/test/java/com/google/cloud/UnmanagedDependencyCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import static org.junit.Assert.assertTrue;
+
+import com.google.cloud.tools.opensource.dependencies.MavenRepositoryException;
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import org.eclipse.aether.resolution.ArtifactDescriptorException;
+import org.eclipse.aether.version.InvalidVersionSpecificationException;
+import org.junit.Test;
+
+public class UnmanagedDependencyCheckTest {
```
Comment (suztomo) [inline java-shared-dependencies/unmanaged-dependency-check/src/test/java/com/google/cloud/UnmanagedDependencyCheckTest.java:13]: Ensure this runs as part of CI.

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.eclipse.aether.resolution.ArtifactDescriptorException;
+import org.eclipse.aether.version.InvalidVersionSpecificationException;
+
+public class UnmanagedDependencyCheck {
+
+  public static void main(String[] args)
```
Comment (suztomo) [inline java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:19]: Would you add Javadoc for the arguments?

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      throws MavenRepositoryException, ArtifactDescriptorException, InvalidVersionSpecificationException {
+    checkArgument(args.length == 2, "The length of the inputs should be 2");
+    System.out.println(getUnmanagedDependencies(args[0], args[1]));
```
Comment (suztomo) [inline java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:22]: What are good error message for library developers?

File: java-shared-dependencies/unmanaged-dependency-check/src/test/java/com/google/cloud/UnmanagedDependencyCheckTest.java (hunk 4 above)
Comment (JoeWang1127) [reply java-shared-dependencies/unmanaged-dependency-check/src/test/java/com/google/cloud/UnmanagedDependencyCheckTest.java:13]: done.

File: java-shared-dependencies/unmanaged-dependency-check/pom.xml (hunk 3 above)
Comment (JoeWang1127) [reply java-shared-dependencies/unmanaged-dependency-check/pom.xml]: done.

File: java-shared-dependencies/unmanaged-dependency-check/action.yaml (hunk 2 above)
Comment (JoeWang1127) [reply java-shared-dependencies/unmanaged-dependency-check/action.yaml]: done.

**Comment by suztomo** [comment]
Let's discuss the design of the interface in https://docs.google.com/document/d/1OSdG4KPtClcE4dPZ2mLowSvCaF1uqUqeQPCxtdtL4c4/edit?resourcekey=0-fkJUdAGqpaH6s4hNM1yEeQ&disco=AAABCv6-GDk

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -277,3 +277,26 @@ jobs:
...
+        distribution: temurin
+    - run: mvn -version
+    - name: Unit Tests
+      run: |
+        mvn test --batch-mode --no-transfer-progress
+      working-directory: java-shared-dependencies/unmanaged-dependency-check
+    - name: Install unmanaged-dependency-check
+      run: |
+        mvn install --batch-mode --no-transfer-progress -DskipTests
+      working-directory: java-shared-dependencies/unmanaged-dependency-check
```
Comment (suztomo) [inline .github/workflows/ci.yaml:298]: I don't think "the user" needs to install java-shared-dependencies/unmanaged-dependency-check. Let's discuss https://docs.google.com/document/d/1OSdG4KPtClcE4dPZ2mLowSvCaF1uqUqeQPCxtdtL4c4/edit?resourcekey=0-fkJUdAGqpaH6s4hNM1yEeQ&disco=AAABCv6-GDk

Comment (JoeWang1127) [reply .github/workflows/ci.yaml]: This step is removed.

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java (hunk 5 above)
Comment (JoeWang1127) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:27]: done.

**Comment by suztomo** [comment]
> Unmanaged dependencies, [com.h2database:h2], found in ../../gapic-generator-java-bom/pom.xml.
> Please follow go/add-dependency-to-cloud-sdk-java to add a third-party dependency.

Two improvements:

1. "../../" is wrong.
2.  I read "gapic-generator-java-bom/pom.xml" but I don't see "com.h2database:h2" there. Can I see the artifact ID that declares the unmanaged dependency?

**Comment by JoeWang1127** [comment]
> "../../" is wrong.
This prefix is now removed.

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java [hunk 8]
```diff
@@ -18,11 +18,13 @@
  * A utility class to check unmanaged dependencies in BOM.
  */
 public class UnmanagedDependencyCheck {
+  // regex of handwritten artifacts
+  private final static String handwrittenArtifact = "(com.google.cloud:google-cloud-.*)|(com.google.api.grpc:(grpc|proto)-google-cloud-.*)";
```
Comment (suztomo) [inline java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:22]: "proto-" or "grpc-" is not handwritten artifact. Can you come up with a better name?

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java (hunk 5 above)
Comment (suztomo) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:29]: I think passing the version of Java shared dependencies will be come inconvenient.

Comment (JoeWang1127) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:29]: Users of unmanaged dependency check don't need to pass this parameter.

The version of shared dependencies needs update every two weeks so I think it's fine to set a parameter.

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java (hunk 8 above)
Comment (JoeWang1127) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:22]: How about `downstreamArtifact` as they are artifacts of of downstream repositories?

File: .github/workflows/create_additional_release_tag.yaml [hunk 9]
```diff
@@ -36,3 +36,10 @@ jobs:
           git tag $TAG_NAME
           git push origin $TAG_NAME
         done
+      # Generate a tag for unmanaged dependencies check.
```
Comment (suztomo) [inline .github/workflows/create_additional_release_tag.yaml:39]: Indent

File: .github/workflows/create_additional_release_tag.yaml [hunk 10]
```diff
@@ -36,3 +36,10 @@ jobs:
...
+      # Use fixed tag so that checks in handwritten libraries do not need to
+      # update the version.
+        FIXED_CHECK_TAG="unmanaged-dependencies-check-latest"
```
Comment (suztomo) [inline .github/workflows/create_additional_release_tag.yaml:42]:
```suggestion
        CHECK_LATEST_TAG="unmanaged-dependencies-check-latest"
```

The tag is not fixed. It's updated every time we make a release of sdk-platform-java.

File: .github/workflows/create_additional_release_tag.yaml [hunk 11]
```diff
@@ -36,3 +36,10 @@ jobs:
...
+        git tag ${FIXED_CHECK_TAG}
+        git push origin ${FIXED_CHECK_TAG}
```
Comment (suztomo) [inline .github/workflows/create_additional_release_tag.yaml:44]: I think you need to add "-f" to update an existing tag.

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java (hunk 6 above)
Comment (suztomo) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:31]: Idea. "This pull request seems to be adding a new third-party dependency com.h2.xxx:h2 among the artifacts listed in BOM com.google.cloud:bigtable-bom. Please see go/cloud-sdk-java-dependency-governance. "

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  private static Set<String> getManagedDependencies(String projectBomPath)
+      throws MavenRepositoryException, InvalidVersionSpecificationException {
+    return getManagedDependenciesFromBom(Bom.readBom(Paths.get(projectBomPath)));
+  }
+
+  private static Set<String> getManagedDependenciesFromBom(Bom bom)
+      throws InvalidVersionSpecificationException {
+    Set<String> res = new HashSet<>();
+    new ClassPathBuilder()
+        .resolve(bom.getManagedDependencies(), true, DependencyMediation.MAVEN)
```
Comment (suztomo) [inline java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:67]: The `resolve()` method returns `ClassPathResult`. This has a convenient method `getDependencyPaths` that tells which artifact actually declares the ClassPathEntry.

```
  /**
   * Returns all paths to the class path entry or 
   * an empty list if the entry is not in the class path.
   */
  public ImmutableList<DependencyPath> getDependencyPaths(ClassPathEntry entry) {
```

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        .getClassPath()
+        .forEach(
+            classPath -> {
+              String coordinate = classPath.toString();
```
Comment (suztomo) [inline java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:71]: Avoid relying on toString(). Use `public Artifact getArtifact()` and read the fields of `Artifact`.

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.eclipse.aether.version.InvalidVersionSpecificationException;
+
+/**
+ * A utility class to check unmanaged dependencies in BOM.
+ */
+public class UnmanagedDependencyCheck {
```
Comment (suztomo) [inline java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:19]: remove public

Comment (JoeWang1127) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java]: done

File: .github/workflows/create_additional_release_tag.yaml (hunk 11 above)
Comment (JoeWang1127) [reply .github/workflows/create_additional_release_tag.yaml]: done

File: .github/workflows/create_additional_release_tag.yaml (hunk 10 above)
Comment (JoeWang1127) [reply .github/workflows/create_additional_release_tag.yaml]: done

File: .github/workflows/create_additional_release_tag.yaml (hunk 9 above)
Comment (JoeWang1127) [reply .github/workflows/create_additional_release_tag.yaml]: done

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java (hunk 13 above)
Comment (JoeWang1127) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java]: done

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java (hunk 12 above)
Comment (JoeWang1127) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:68]: As discussed, this enhancement will go to the next PR.

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+  // regex of handwritten artifacts
+  private final static String downstreamArtifact = "(com.google.cloud:google-cloud-.*)|(com.google.api.grpc:(grpc|proto)-google-cloud-.*)";
+
+
+  /**
+   * @param args An array with two elements.<p> The first string is the version of Java shared
```
Comment (suztomo) [inline java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:26]: Outdated

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java (hunk 14 above)
Comment (suztomo) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:20]: Strange. But fine.

**Review by suztomo — APPROVED** [review]

File: java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java (hunk 15 above)
Comment (suztomo) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java:26]: > The first string is the version of Java shared dependencies.

Comment (JoeWang1127) [reply java-shared-dependencies/unmanaged-dependency-check/src/main/java/com/google/cloud/UnmanagedDependencyCheck.java]: done

**Pull request merged**